provider/equinix/environ.go

```diff
@@ -78,31 +81,100 @@ func (e *environ) Config() *config.Config {
 }
 
 func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
-	return nil, errors.NewNotImplemented(nil, "not implemented")
+	validator := constraints.NewValidator()
+	validator.RegisterUnsupported([]string{constraints.CpuPower, constraints.VirtType})
+	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
 }
 
 func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
-	return nil, errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-is-controller": "true", "juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return nil, err
+	}
+	instanceIDs := make([]instance.Id, len(insts))
+	for i, inst := range insts {
+		instanceIDs[i] = inst.Id()
+	}
+	return instanceIDs, nil
 }
 
 func (e *environ) Create(ctx context.ProviderCallContext, args environs.CreateParams) error {
 	return nil
 }
 
 func (e *environ) Destroy(ctx context.ProviderCallContext) error {
-	return errors.NewNotImplemented(nil, "not implemented")
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, inst := range insts {
+		if _, err = e.equinixClient.Devices.Delete(string(inst.Id()), true); err != nil {
```

Comment:
Should we immediately bail out on error or release as many devices as we can first and then return back the errors?

Reply:
This sounds like a concern that who designed this project should answer. How other providers do that and do you have a way to keep this behavior consistent across providers?

Reply:
IIRC, the other providers exit early so I guess you can leave this as-is for now.